### DISCUSSION TITLE
Upgrade from SQLAlchemy 1.3 => 1.4 - issue with fork() and engine.dispose()

### DISCUSSION THREAD

**Discussion by stephan1373** [sqlalchemy/sqlalchemy#7412] [Usage Questions]
Hello,  this is our second issue related to a recent upgrade to SQLAlchemy 1.4.   

Our application stack uses flask + gunicorn with: 
-  preload_app=True (loads the app once, then forks it for N number of gunicorn worker child processes) 
- With SQLAlchemy 1.3, we followed the advice labeled [Using Connection Pools with Multiprocessing or os.fork()](https://docs.sqlalchemy.org/en/13/core/pooling.html#pooling-multiprocessing) 
- By following this advice, we avoided issues like `sqlalchemy.exc.OperationalError: (MySQLdb._exceptions.OperationalError) (2006, 'Server has gone away')` 

With SQLAlchemy 1.4,:
- we reviewed the same advice ([Using Connection Pools with Multiprocessing or os.fork()](https://docs.sqlalchemy.org/en/14/core/pooling.html#pooling-multiprocessing)) (seems to be the same) , and 
- Unfortunately, we are now seeing exceptions that still look a lot like there are a direct result of calling fork() - including the "Server has gone away" exception, above. 

Our current work-around: is to set `preload_app=False` to gunicorn.py so that each process starts independently (i.e., without a fork()).  However, we would strongly prefer to use the same approach as 1.3 and fix whatever is happening after we fork().   

Is there something else we need to call or setup in addition to the existing advice already listed?   Those details are described in the link above, but I can confirm that we are:
- calling engine.dispose()
- using the recommended event handlers (for connect, checkout (connection) events)

I'll add some log files here if that helps (although they will also have some of our application code, too).   If we set preload_app=False, the application appears to work fine.  

(The advantage of preload_app=True is that it allows us to perform initialization more cleanly without contention (especially on an uninitialized database))

Also if it helps, we had another [question recently here ](https://github.com/sqlalchemy/sqlalchemy/discussions/7359) related to the upgrade - this should have information about our specific software stack.

[logfile1.txt](https://github.com/sqlalchemy/sqlalchemy/files/7670817/logfile1.txt)
[logfile2.txt](https://github.com/sqlalchemy/sqlalchemy/files/7670818/logfile2.txt)

**Comment by zzzeek** [comment]
first thing -

can you confirm switching to NullPool removes the problem entirely?  that would at least isolate this down to an issue of connections coming across the process boundary.

**Reply by zzzeek** [reply]
also if you could confirm you normally use QueuePool (that is, you arent using a pool like StaticPool or something)

**Reply by zzzeek** [reply]
I notice you work with the eNMS application, where we also fixed some other issues, is this involving eNMS ?  is #7359 the other issue you're referring towards?  EDIT nevermind I see you mentioned it, OK

**Comment by zzzeek** [comment]
regarding https://docs.sqlalchemy.org/en/13/core/pooling.html#pooling-multiprocessing , the only method you need to use is to be calling engine.dispose() at the start of the fork.     the third option with the event handler im not sure that we have tested under 1.4, it should work fine, however, .dispose() itself should be sufficient.   nothing has really changed with how dispose() works, it not only tells the pool to explicitly close all of its pooled connections, it entirely replaces the pool with a new one as well, so there's really no chance of a connection from the pre-disposed engine being produced by the post-disposed engine.

**Reply by zzzeek** [reply]
I'm assuming your approach is this: https://github.com/stephan1373/eNMS/blob/075bb5d590c873c35155b5ef5884ac8125107f74/gunicorn.py#L17 

I would recommend turning on echo_pool=True in create_engine(), making sure this dispose happens at the right time.

if you are able to reproduce the problem in an isolated server, such as by using a script to send API requests concurrently, that would be very helpful such that we could turn on more aggressive logging (less approrpriate for production loads) and see what's happening.

**Reply by stephan1373** [reply]
We have some more information and we have tested some additional scenarios:
- Using the NullPool did not remove the issue (we can see the pool getting disposed)
- Switching from MySQLdb to PyMySQL - still had similar DB issues (in one run, we got a sequence number error)
- Just **rolling back SQL Alchemy 1.3.24** did not remove DB-related errors (this was unexpected!)

To verify, we took an older system that did not appear to have these issues 
- And confirmed that the old system seems to work fine.

So, it looks like something else is still going on.   We're going to reset and do some more testing again tomorrow, but wanted to share what we think we saw today (repeated observations are planned).

**Reply by zzzeek** [reply]
OK yes, there have been changes to the pool in 1.4, but not regarding dispose(); that's a very blunt tool and nothing changed there.

Before I just saw this I wrote a quick POC of a program that uses multiprocessing and reproduces the very common "MySQL server has gone away" errors immediately and profusely if the line to engine.dispose() is commented out; restoring the line and the script runs for many minutes until the end with no errors.

```py
from multiprocessing import Process
import os
import time

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session



Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String(50))
    bs = relationship("B")


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String(50))



def insert_data():
    # comment out this line and the program fails immediately
    # every time with many "MySQL server has gone away"
    # and other concurrency errors
    e.dispose()

    for x in range(300):
        with Session(e) as sess:
            sess.add_all([
                A(data="some data", bs=[B(), B(), B()])
                for i in range(30)
            ])
            print(f"process {os.getpid()} inserted 30 records")
            sess.commit()
        time.sleep(.05)

if __name__ == '__main__':
    e = create_engine("mysql://scott:tiger@localhost/test")
    Base.metadata.drop_all(e)
    Base.metadata.create_all(e)

    procs =[Process(target=insert_data) for j in range(20)]
    for p in procs:
        p.start()
    for p in procs:
        p.join()
```
